### DISCUSSION TITLE
Usecase: Connecting to MetaMask

### DISCUSSION THREAD

**Discussion by numpde** [idom-team/idom#978] [Question] [answered]
Consider a website with

o) a front-end that uses [MetaMask](https://img.crx4chrome.com/63/81/41/nkbihfbeogaeaoehlefnkodbefgpgknn-screenshot.webp) to digitally sign transactions;
o) a back-end (reactpy-django) that talks to the blockchain, etc.

Below is a self-contained sample app to explain some basic dynamics (you'll need MetaMask or another "wallet").

How "natural" would it be to do the same using ReactPy? I expect not very.

But, what would be a good way to have this app talk to ReactPy? 

I hope the gist of the question is clear... 

==

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Wallet connection test</title>
    <style>
        .fade-in {
            opacity: 0;
            transition: opacity 0.5s ease-out;
        }

        .fade-in.show {
            opacity: 1;
        }
    </style>
</head>
<body>
<div id="root"></div>

<script src="https://cdn.jsdelivr.net/gh/ethereum/web3.js/dist/web3.min.js" crossorigin="anonymous"></script>
<script src="https://unpkg.com/react@18/umd/react.development.js" crossorigin="anonymous"></script>
<script src="https://unpkg.com/react-dom@18/umd/react-dom.development.js" crossorigin="anonymous"></script>
<script src="https://unpkg.com/babel-standalone@6/babel.min.js" crossorigin="anonymous"></script>

<script type="text/babel">
    const DEFAULT_NETWORK = '0x89';

    class MyApp extends React.Component {
        constructor(props) {
            super(props);
            this.state = {
                account: '',
                network: '',
                isConnected: undefined,
            };
        }

        componentDidMount = async () => {
            const network = await this.getCurrentNetwork();
            const account = await this.getCurrentAccount();

            this.setState({
                network,
                account,
                isConnected: !!account,
            });

            ethereum.on('accountsChanged', this.handleAccountsChanged);
            ethereum.on('chainChanged', this.handleChainChanged);
        }

        componentWillUnmount() {
            ethereum.off('accountsChanged', this.handleAccountsChanged);
            ethereum.off('chainChanged', this.handleChainChanged);
        }

        handleAccountsChanged = async (accounts) => {
            if (accounts) {
                this.setState({
                    account: accounts[0],
                    isConnected: true,
                });
            } else {
                this.setState({
                    account: '',
                    isConnected: false,
                });
            }
        }

        handleChainChanged = async (chainId) => {
            this.setState({
                network: chainId,
            });
        }

        getCurrentNetwork = async () => {
            return await ethereum.request({method: 'eth_chainId'});
        }

        getCurrentAccount = async () => {
            const accounts = await ethereum.request({method: 'eth_accounts'});
            return accounts[0];
        }

        connect = async () => {
            const accounts = await ethereum.request({method: 'eth_requestAccounts'});
            await this.handleAccountsChanged(accounts);
        }

        switchToDefaultNetwork = async () => {
            await ethereum.request({
                method: 'wallet_switchEthereumChain',
                params: [{chainId: DEFAULT_NETWORK}],
            });
        }

        render() {
            const {isConnected, account, network} = this.state;

            return (
                <div className={`fade-in ${(isConnected !== undefined) ? 'show' : ''}`}>
                    <button onClick={this.connect} disabled={isConnected}>
                        {isConnected ? 'Connected to wallet' : 'Connect to wallet...'}
                    </button>
                    <div>Account: {account || 'Not connected'}</div>
                    <div>Network: {network || 'Unknown'}</div>
                    {
                        network && (network !== DEFAULT_NETWORK) && (
                            <div>
                                <p>Not on the default network.</p>
                                <button onClick={this.switchToDefaultNetwork}>Switch to default</button>
                            </div>
                        )
                    }
                </div>
            );
        }
    }

    const root = ReactDOM.createRoot(document.getElementById('root'));
    root.render(<MyApp/>);
</script>

</body>
</html>

```

**Comment by rmorshea** [comment, marked as answer]
If a majority of the work _must_ be done client-side, then ReactPy may not be the best option. If, on the other hand, a majority of the work can be done server-side, and you just need complex client-side logic in a few places, using a [custom JS component](https://reactpy.dev/docs/guides/escape-hatches/javascript-components.html#custom-javascript-components) or two could be a reasonable solution. I'm happy to explain how to do this in more detail if the docs are not sufficient.

For your particular case, doing a quick look around, it does seem like there are [Python libraries](https://web3py.readthedocs.io/en/stable/) for working with Etherum. So on it's face, it seems like you could move some of this logic server-side. I just don't know enough about what you're trying to do specifically to say whether or not what you're trying to do can/should be done mostly server-side.

Eventually I can see a world in which ReactPy can supply client-side components written in pure Python via something like Pyodide, but that's a long way off at the moment.

**Reply by numpde** [reply]
Custom JS component might be enough but I don't see from the docs how to communicate both ways. Coming back to "MyApp" above, a viable minimum would be to send to it a message-to-sign from a ReactPy component elsewhere on the page and receive the result. That component would do the blockchain-related business logic server-side. 

Or, in terms of the [super simple chart](https://reactpy.dev/docs/guides/escape-hatches/javascript-components.html#custom-javascript-components), how to handle events? Let's say a user clicks on the chart to make a new datapoint.

**Reply by rmorshea** [reply]
So the `props` passed to the JS `function SuperSimpleChart(props)` will contain both static props (e.g. numbers, strings) as well as callbacks that were passed to the associated server-side component. You can see a real-world example in [`reactpy-router`](https://github.com/reactive-python/reactpy-router) where an `onClick` handler is passed [from Python](https://github.com/reactive-python/reactpy-router/blob/main/reactpy_router/core.py#L71), to the [JS component](https://github.com/reactive-python/reactpy-router/blob/main/js/src/index.js#L18).

In short, all you have to do is specify a callable attribute and ReactPy will do the necessary work to link it up with its client-side equivalent. Then any positional `...args` passed via `props.yourCallback(...args)` will be handed off to the server-side `your_callback(*args)` function so long as `...args` contains JSON serializable values.

**Reply by rmorshea** [reply]
I do think it's worth considering what parts of your app _have_ to be client-side though. At a glance, I don't really see anything about your use case that couldn't be done using Python libraries server-side. Perhaps there are security implications I'm missing though.

**Reply by rmorshea** [reply]
Also, I'm working on renaming the custom JS template repo [right now](https://github.com/idom-team/idom-react-component-cookiecutter/pull/12). Should be done shortly.

**Reply by numpde** [reply]
> I do think it's worth considering what parts of your app _have_ to be client-side though. At a glance, I don't really see anything about your use case that couldn't be done using Python libraries server-side. Perhaps there are security implications I'm missing though.

The cryptographic signature using the wallet browser plugin (in the above case `ethereum`) is client-side. I didn't spell out this part in the MyApp. The alternative would be passing a private key to the server, which is usually a no-go.

Thanks for the answers.
